src/devices/bus/ti99/gromport/cartridges.cpp

```diff
@@ -1571,178 +1522,68 @@ void ti99_cartridge_device::rpk::close()
 {
-	// find the length attribute
-	std::string const *const length_string = ram_resource_node->get_attribute_string_ptr("length");
-	if (length_string == nullptr) throw rpk_exception(RPK_MISSING_RAM_LENGTH);
-
-	// parse it
-	unsigned int length;
-	char suffix = '\0';
-	sscanf(length_string->c_str(), "%u%c", &length, &suffix);
-	switch(tolower(suffix))
-	{
-		case 'k': // kilobytes
-			length *= 1024;
-			break;
-
-		case 'm':
-			/* megabytes */
-			length *= 1024*1024;
-			break;
-
-		case '\0':
-			break;
-
-		default:  // failed
-			throw rpk_exception(RPK_INVALID_RAM_SPEC);
-	}
-
 	// Allocate memory for this resource
-	std::unique_ptr<uint8_t []> contents;
-	try { contents = make_unique_clear<uint8_t []>(length); }
-	catch (std::bad_alloc const &) { throw rpk_exception(RPK_OUT_OF_MEMORY); }
+	std::vector<uint8_t> contents;
+	contents.resize(socket.length());
+	std::fill(contents.begin(), contents.end(), 0);
 
-	LOGMASKED(LOG_RPK, "[RPK handler] Allocating RAM buffer (%d bytes) for socket '%s'\n", length, socketname);
+	LOGMASKED(LOG_RPK, "[RPK handler] Allocating RAM buffer (%d bytes) for socket '%s'\n", socket.length(), socket.id());
 
 	// That's it for pure RAM. Now check whether the RAM is "persistent", i.e. NVRAM.
 	// In that case we must load it from the NVRAM directory.
 	// The file name is given in the RPK file; the subdirectory is the system name.
-	std::string const *const ram_type = ram_resource_node->get_attribute_string_ptr("type");
 	std::string ram_pname;
-	if (ram_type != nullptr)
+	if (socket.type() == util::rpk_socket::socket_type::PERSISTENT_RAM)
 	{
-		if (*ram_type == "persistent")
-		{
-			// Get the file name (required if persistent)
-			std::string const *const ram_filename = ram_resource_node->get_attribute_string_ptr("file");
-			if (ram_filename==nullptr)
-				throw rpk_exception(RPK_INVALID_RAM_SPEC, "<ram type='persistent'> must have a 'file' attribute");
-
-			ram_pname = std::string(system_name).append(PATH_SEPARATOR).append(*ram_filename);
-			// load, and fill rest with 00
-			LOGMASKED(LOG_RPK, "[RPK handler] Loading NVRAM contents from '%s'\n", ram_pname.c_str());
-
-			// Load the NVRAM contents
-			if (!contents || (length <= 0))
-				throw emu_fatalerror("ti99_cartridge_device::rpk_reader::load_ram_resource: Buffer is null or length is 0");
-
-			// try to open the battery file and read it if possible
-			emu_file file(options.nvram_directory(), OPEN_FLAG_READ);
-			std::error_condition const filerr = file.open(ram_pname);
-			int bytes_read = 0;
-			if (!filerr)
-				bytes_read = file.read(contents.get(), length);
+		ram_pname = std::string(system_name).append(PATH_SEPARATOR).append(socket.filename());
+		// load, and fill rest with 00
+		LOGMASKED(LOG_RPK, "[RPK handler] Loading NVRAM contents from '%s'\n", ram_pname.c_str());
```

Comment:
You shouldn’t be using `.c_str()` when passing strings to our formatting functions.  They can all deal with most string-like things.